### PR TITLE
Add new darc commands `vmr backflow` and `vmr forwardflow`

### PR DESCRIPTION
**Pull request by premun**
These are new commands for flowing code between the VMR and repositories. The PR also contains E2E tests that test the basic code flow scenarios.

The commands end up commiting the flow changes in a local branch so all works locally only.

There are TODOs left in the code intentionally to identify places/edge cases where we will need to come up with some graceful error handling but it's unclear at the moment what the outcome should be. This will become clear once we integrate this into a service.

https://github.com/dotnet/arcade-services/issues/2996

### Release Note Category
- [x] Feature changes/additions 
- [ ] Bug fixes
- [ ] Internal Infrastructure Improvements

### Release Note Description
Added new darc commands `vmr backflow` and `vmr forwardflow`

### CODE REVIEW

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+        SourceMapping mapping,
+        string shaToFlow,
+        NativePath repoPath,
+        Codeflow lastFlow,
+        bool discardPatches,
+        CancellationToken cancellationToken)
+    {
+        var isBackflow = lastFlow is Backflow;
+        var shortShas = $"{Commit.GetShortSha(lastFlow.SourceSha)}-{Commit.GetShortSha(shaToFlow)}";
+        var branchName = $"codeflow/{lastFlow.GetType().Name.ToLower()}/{shortShas}";
```
Comment (tkapin) [inline src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs:111]: Strong recommendation to avoid using reflection unless necessary (for many reasons). Introduce a `string Type` property on the CodeFlow and its  descendants and use it instead.

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+        var targetRepo = isBackflow ? repoPath : _vmrInfo.VmrPath;
```
Comment (tkapin) [inline src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs:112]: This logic could / should be expressed in the ForwardFlow / BackFlow classes if they are enriched by the Repo and VMR paths. This is a second case where I think it would be beneficial.

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+        var submoduleExclusions = _sourceManifest.Submodules
```
Comment (tkapin) [inline src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs:113]: How does this work? Why we don't exclude all submodules? I think an explanatory comment would be useful here.

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+            {
+                await _vmrPatchHandler.ApplyPatch(patch, targetRepo, cancellationToken);
+
+                if (discardPatches)
+                {
+                    _fileSystem.DeleteFile(patch.Path);
+                }
+            }
+        }
+        catch (Exception e) when (e.Message.Contains("Failed to apply the patch"))
```
Comment (tkapin) [inline src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs:169]: Consider: introduce a dedicated exception type for this rather than relying on the exception message.

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+
+        try
+        {
+            foreach (VmrIngestionPatch patch in patches)
+            {
+                await _vmrPatchHandler.ApplyPatch(patch, targetRepo, cancellationToken);
+
+                if (discardPatches)
+                {
+                    _fileSystem.DeleteFile(patch.Path);
```
Comment (tkapin) [inline src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs:165]: Consider: pass the discard to the ApplyPatch() method and do the cleanup there. It will save the if and would logically fit there.

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+                }
+            }
+        }
+
+        var commitMessage = $"""
+            [{(isBackflow ? "VMR" : mapping.Name)}] Codeflow {shortShas}
```
Comment (tkapin) [inline src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs:207]: Another place where `{(isBackflow ? "VMR" : mapping.Name)}` would be benefitial to have as a method of the code flow (presuming it is enriched by information about the repo and vmr, not just the shas).

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+        Codeflow lastFlow,
+        bool discardPatches,
+        CancellationToken cancellationToken)
+    {
+        var shortShas = $"{Commit.GetShortSha(lastFlow.SourceSha)}-{Commit.GetShortSha(shaToFlow)}";
+        var patchName = _vmrInfo.TmpPath / $"{mapping.Name}-backflow-{shortShas}.patch";
+
+        await _localGitClient.CheckoutAsync(targetRepo, lastFlow.SourceSha);
+
+        var branchName = $"codeflow/{lastFlow.GetType().Name.ToLower()}/{shortShas}";
```
Comment (tkapin) [inline src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs:233]: Reflection, commented on another place.

**Review by tkapin — COMMENTED** [review]
Overall, my main observation is that it would be beneficial to extend the CodeFlow and its derived classes with additional information (mainly the VMR and repo) and use it consistently (several places in the code marked where it would fit naturally).

Second, more fundamental comment (possibly for synchronous discussion) is about introducing a type representing a git repo, rather than expressing the repo as string. The main argument for that is that a git repo is a fundamental entity in this domain and many operations are done naturally on the repo. I understand the intent to decouple it from the git client, but I think the code would flow / read more natural with a proper repo type.

File: src/Microsoft.DotNet.Darc/Darc/Options/VirtualMonoRepo/CodeFlowCommandLineOptions.cs [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using CommandLine;
+
+namespace Microsoft.DotNet.Darc.Options.VirtualMonoRepo;
+
+internal abstract class CodeFlowCommandLineOptions : VmrCommandLineOptions
+{
+    [Option("additional-remotes", Required = false, HelpText =
```
[the remark this answers is not in the document]
Comment (premun) [reply src/Microsoft.DotNet.Darc/Darc/Options/VirtualMonoRepo/CodeFlowCommandLineOptions.cs:11]: We need to be able to provide a remote in almost all scenarios during PR builds where we point to a local clone.
We also need to be able to specify custom fork when manually synchronizing changes from it.

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+        await _localGitClient.CommitAsync(targetRepo, commitMessage, false, cancellationToken: cancellationToken);
+        await _localGitClient.ResetWorkingTree(targetRepo);
+
+        return branchName;
+    }
+
+    private async Task UpdateVersionDetailsXml(NativePath repoPath, string currentVmrSha, CancellationToken cancellationToken)
+    {
+        // TODO: Do a proper full update of all dependencies, not just V.D.xml
+        var versionDetailsXml = DependencyFileManager.GetXmlDocument(await _fileSystem.ReadAllTextAsync(repoPath / VersionFiles.VersionDetailsXml));
```
[the remark this answers is not in the document]
Comment (premun) [reply src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs:310]: Yes, this is correct. It's expected that this will be re-done once we have a full list of dependencies. For that, we have this functionality already encapsulated in `DependencyFileManager` and no XMLs will be handled here.
There is a TODO above for that but we can't implement that yet as we don't have VMR BAR builds yet.

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrCodeflower.cs [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+
+        if (!string.IsNullOrEmpty(result.StandardError))
+        {
+            result.ThrowIfFailed($"Failed to determine which commit of {repoPath} is older ({parent}, {ancestor})");
+        }
+
+        return result.ExitCode == 0;
+    }
+
+    protected abstract record Codeflow(string SourceSha, string TargetSha)
```
[the remark this answers is not in the document]
Comment (premun) [reply src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrCodeflower.cs:268]: > What I don't understand though is how the Codeflow properties RepoSha and VmrSha are mapped to the primary parameters of SourceSha and TargetSha

The child classes implement these abstract properties by defining those via the primary constructor (which makes the parameters public properties).

> I understand it was intentional, but it would make sense to me to define ForwardFlow as ForwardFlow(string RepoSha, string VmrSha) so that the reading direction follows the flow.

A good idea, I will change this so that the order of the parameters represents the direction of the flow.

> These three records are very confusing, especially the naming of the arguments and properties.

This actually helps a lot in the usage where for creating these, it's just simpler to match the parameters by name (`vmrSha`, `repoSha`) and not care about the order of things since we already know which flow we're creating.
But later, in common parts for back and forward flow, it's more readable to use `Source` and `Target` as it simplifies the reading. So that we don't need to do `isBackflow ? flow.RepoSha : flow.VmrSha` all the time.

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrCodeflower.cs [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+        ForwardFlow lastForwardFlow = await GetLastForwardFlow(mapping.Name);
+        Backflow? lastBackflow = await GetLastBackflow(repoPath);
+
+        if (lastBackflow is null)
+        {
+            return lastForwardFlow;
+        }
+
+        string backwardSha, forwardSha;
+        NativePath sourceRepo;
```
[the remark this answers is not in the document]
Comment (premun) [reply src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrCodeflower.cs:181]: Created https://github.com/dotnet/arcade-services/issues/3155 as a follow-up

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrCodeflower.cs [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+        else
+        {
+            (backwardSha, forwardSha) = (lastBackflow.RepoSha, lastForwardFlow.RepoSha);
+            sourceRepo = repoPath;
+        }
+
+        string objectType1 = await _localGitClient.GetObjectTypeAsync(sourceRepo, backwardSha);
+        string objectType2 = await _localGitClient.GetObjectTypeAsync(sourceRepo, forwardSha);
+
+        if (objectType1 != "commit" || objectType2 != "commit")
```
[the remark this answers is not in the document]
Comment (premun) [reply src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrCodeflower.cs:196]: Done, returning an enum now

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrCodeflower.cs [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+    }
+
+    /// <summary>
+    /// Compares 2 git commits and returns true if the first one is an ancestor of the second one.
+    /// </summary>
+    private async Task<bool> IsAncestorCommit(NativePath repoPath, string parent, string ancestor)
+    {
+        var result = await _processManager.ExecuteGit(repoPath, ["merge-base", "--is-ancestor", parent, ancestor]);
+
+        if (!string.IsNullOrEmpty(result.StandardError))
```
[the remark this answers is not in the document]
Comment (premun) [reply src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrCodeflower.cs:260]: Done

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrCodeflower.cs [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+        return isBackwardOlder ? lastForwardFlow : lastBackflow;
+    }
+
+    /// <summary>
+    /// Finds the last backflow between a repo and a VMR.
+    /// </summary>
+    private async Task<Backflow?> GetLastBackflow(NativePath repoPath)
+    {
+        // Last backflow SHA comes from Version.Details.xml in the repo
+        var content = await _fileSystem.ReadAllTextAsync(repoPath / VersionFiles.VersionDetailsXml);
```
[the remark this answers is not in the document]
Comment (premun) [reply src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrCodeflower.cs:221]: Done

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrForwardFlower.cs [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+        if (shaToFlow is null)
+        {
+            shaToFlow = await _localGitClient.GetShaForRefAsync(sourceRepo, Constants.HEAD);
+        }
+        else
+        {
+            await _localGitClient.CheckoutAsync(sourceRepo, shaToFlow);
+        }
+
+        return await FlowCodeAsync(
```
[the remark this answers is not in the document]
Comment (premun) [reply src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrForwardFlower.cs:76]: Done

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs (hunk 2 above)
Comment (premun) [reply src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs:112]: Ah, this logic can actually go away because we know we are doing a backflow

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs (hunk 3 above)
Comment (premun) [reply src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs:113]: Added a comment

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs (hunk 6 above)
Comment (premun) [reply src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs:207]: Removed, it's actually obvious this will be always true

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs (hunk 7 above)
Comment (premun) [reply src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrBackflower.cs:233]: Done

**Comment by premun** [comment]
@tkapin can you have another look? I think I resolved all of the comments

File: src/Microsoft.DotNet.Darc/DarcLib/LocalGitClient.cs [hunk 16]
```diff
@@ -194,7 +194,14 @@ public async Task<string> GetObjectTypeAsync(string repoPath, string objectSha)
         var result = await _processManager.ExecuteGit(repoPath, args);
         result.ThrowIfFailed($"Failed to find object {objectSha} in {repoPath}");
 
-        return result.StandardOutput.Trim();
+        return result.StandardOutput.Trim() switch
```
Comment (tkapin) [inline src/Microsoft.DotNet.Darc/DarcLib/LocalGitClient.cs:197]: Nice!

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrCodeflower.cs [hunk 17]
```diff
@@ -270,9 +269,19 @@ protected abstract record Codeflow(string SourceSha, string TargetSha)
         public abstract string RepoSha { get; init; }
 
         public abstract string VmrSha { get; init; }
+
+        public string GetBranchName() => $"codeflow/{Name}/{Commit.GetShortSha(SourceSha)}-{Commit.GetShortSha(TargetSha)}";
+
+        public abstract string Name { get; }
```
Comment (tkapin) [inline src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrCodeflower.cs:275]: Consider: use Enum for this to make clear what the possible values are.

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrCodeflower.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (tkapin) [reply src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrCodeflower.cs:267]: > But later, in common parts for back and forward flow, it's more readable to use Source and Target as it simplifies the reading. So that we don't need to do isBackflow ? flow.RepoSha : flow.VmrSha all the time.

Agree on this.

**Review by tkapin — APPROVED** [review]

File: src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrCodeflower.cs (hunk 17 above)
Comment (premun) [reply src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/VmrCodeflower.cs:275]: I introduced this so that we can avoid the `flow is BackFlow ? "backward" : "forward"` in quite a few places. The enum would have to be named lowercased or we would have to keep lowercasing it. Is it worth it?
